### ISSUE TITLE
[Question] Jupyterhub leveraging a "Serverless" architecture?

### ISSUE THREAD

**Issue by scottx611x** [jupyterhub/jupyterhub#970]
Hi all, I'm a part of an upcoming hackathon in a month and had the lofty idea of somehow making Jupyterhub work nicely with a tool called [Zappa](https://github.com/Miserlou/Zappa).

In short, this would allow for extremely cheap Jupyterhub hosting with high horizontal scalability.

There are obviously some limitations with this approach, with a small list of ones known by me being:
- Zappa requires a WSGI-based application to deploy
- Each notebook itself requires a separate kernel process to run upon
- User data needs to persist .... somewhere

Currently my colleague is running a decently sized Jupyterhub on an EC2 instance at a cost of $1000 a month, so if possible it seems like a very worthwhile project. 

I'm open to suggestions of more sustainable solutions than the one mentioned above. I've seen solutions utilizing Docker Swarm that looked pretty slick, but those still seemed to only be able to serve temporary notebook instances. Also please keep in mind that within the hackathon context we're being pushed to make something that **doesn't already exist**!

**TL;DR:** What are the limitations of running Jupyterhub behind one of these newfangled "serverless" architectures?

**Comment by barrachri** [comment]
AWS Lambda has (or had) huge limitations....

5 minutes, max duration of the function
You can just fit 50MB of zipped packages.

From my opinion (because I tried) it doesn't make any sense to have jupyter there.

**Comment by scottx611x** [comment]
@barrachri Thanks for the quick response. 

I believe that this Zappa tool got around the 50MB limit by having some setting to just upload a small handler to Lambda and load your actual project from S3 at runtime.

The 5 minute runtime is definitely a valid concern, but for my lab's use case I don't envision many people needing more than 5 minutes to run their Python code.

If theres any chance at an implementation (even if its deemed super hacky) I think it would cool to see it be possible, and would appreciate any advice on getting such a thing to work.

**Comment by barrachri** [comment]
Hey @scottx611x hacking is always cool ! :D

I have some experience with AWS Lambda cause I tried to deploy machine learning model using it.

Related to jupyterhub you also have the proxy that runs with the Hub.

**Comment by scottx611x** [comment]
@barrachri  Indeed it is :) I'm going to do a bunch of reading in the next month. Where heres a will theres a way right? 

If you wouldn't mind keeping this open for a months time in case anyone else has opinions or ideas that would be awesome, thanks!

**Comment by minrk** [comment]
JupyterHub's architecture doesn't seem like it will be a good fit for a 'serverless' architecture, mainly because it's based on tornado async instead of wsgi (tornado can play nice with wsgi, *as long as you don't use async*). Wrapping a heavily async tornado app in wsgi seems like a difficult task to accomplish well. Anything is possible, though.

That said, all of JupyterHub's 'state' does reside in a database. This means that it may technically be possible to run handlers separately with connections to this database. It would mean instantiating the entire application for each handler, which doesn't seem like it would have great performance.

It's probably the single-user servers that are hardest to fit, because they don't persist state to a database, their states (running kernels, etc.) only reside in memory. It's possible that a custom KernelManager and SessionManager could persist and resume this runtime state, though.

Perhaps the hardest part is going to be the fact that JupyterHub and the notebook in general are designed to expose persistent computational resources, which is a bit at odds with serverless design.

**Comment by scottx611x** [comment]
@minrk Thanks! I really appreciate the insight. I figured there would be some hurdles here, but didn't realize they'd be so tall.

**Comment by willingc** [comment]
@minrk and others, I'm going to go ahead and close this question as resolved.
